src/libs/registry/docset.cpp

```diff
@@ -607,3 +612,178 @@ QString Docset::parseSymbolType(const QString &str)
+                        break;
+                    first = false;
+                }
+                if (i != 0) {
+                    distance += 1;
+                    if (distance > 8) {  // optimization #2: too large distance between found chars
+                        break;
+                    }
+                }
+            }
+        }
+
+        if (!found) {
+            // end of haystack, char not found
+            *start = -1;
+            return;
+        }
+    }
+}
+
+static int scoreExact(int matchIndex, int matchLen, const char* value, int valueLen)
+{
+    int score = 100;
+    const char DOT = '.';
+    // Remove one point for each unmatched character.
+    score -= (valueLen - matchLen);
+    if (matchIndex > 0) {
+        if (value[matchIndex - 1] == DOT) {
+            // If the character preceding the query is a dot, assign the same
+            // score as if the query was found at the beginning of the string,
+            // minus one.
+            score += (matchIndex - 1);
+        } else if (matchLen == 1) {
+            // Don't match a single-character query unless it's found at the
+            // beginning of the string or is preceded by a dot.
+            return 0;
+        } else {
+            // (1) Remove one point for each unmatched character up to
+            //     the nearest preceding dot or the beginning of the
+            //     string.
+            // (2) Remove one point for each unmatched character
+            //     following the query.
+            int i = matchIndex - 2;
+            while (i >= 0 && value[i] != DOT) --i;
+            score -= (matchIndex - i) +                     // (1)
+                     (valueLen - matchLen - matchIndex);  // (2)
+        }
+
+        // Remove one point for each dot preceding the query, except for the
+        // one immediately before the query.
+        int separators = 0,
+            i = matchIndex - 2;
+
+        while (i >= 0) {
+            if (value[i] == DOT)
+                ++separators;
+            --i;
+        }
+
+        score -= separators;
+    }
+
+    // Remove five points for each dot following the query.
+    int separators = 0;
+    int i = valueLen - matchLen - matchIndex - 1;
+    while (i >= 0) {
+        if (value[matchIndex + matchLen + i] == DOT) {
+            ++separators;
+        }
+        --i;
+    }
+
+    score -= separators * 5;
+
+    return qMax(1, score);
+}
+
+static int scoreFuzzy(int matchIndex, int matchLen, const char *value)
+{
+    if (matchIndex == 0 || value[matchIndex - 1] == '.') {
+        return qMax(66, 100 - matchLen);
+    } else {
+        if (value[matchLen] == 0) {
+            return qMax(33, 67 - matchLen);
+        } else {
+            return qMax(1, 34 - matchLen);
+        }
+    }
+}
+
+static int scoreFunc(const char *needleOrig, const char* haystackOrig)
+{
+    int haystackLen = 0, needleLen = 0;
+    while (haystackOrig[++haystackLen] != 0);
```

Comment:
This is hard to understand, please replace with a loop with `++haystackLen` inside.